genesis/engine/solvers/rigid/rigid_solver_decomp.py

```diff
@@ -4871,59 +4913,101 @@ def kernel_update_vgeoms(
+    entities_state: array_class.EntitiesState,
+    entities_info: array_class.EntitiesInfo,
+    links_state: array_class.LinksState,
+    geoms_state: array_class.GeomsState,
+    collider_state: array_class.ColliderState,
+    unused__rigid_global_info: array_class.RigidGlobalInfo,
     static_rigid_sim_config: ti.template(),
-):
+    contact_island: ti.template(),  # ContactIsland,
+) -> None:
 
     n_entities = entities_state.hibernated.shape[0]
     _B = entities_state.hibernated.shape[1]
+    ci = contact_island
+
     ti.loop_config(serialize=static_rigid_sim_config.para_level < gs.PARA_LEVEL.PARTIAL)
-    for i_e, i_b in ti.ndrange(n_entities, _B):
-        if (
-            not entities_state.hibernated[i_e, i_b] and entities_info.n_dofs[i_e] > 0
-        ):  # We do not hibernate fixed entity
-            hibernate = True
-            for i_d in range(entities_info.dof_start[i_e], entities_info.dof_end[i_e]):
-                if (
-                    ti.abs(dofs_state.acc[i_d, i_b]) > static_rigid_sim_config.hibernation_thresh_acc
-                    or ti.abs(dofs_state.vel[i_d, i_b]) > static_rigid_sim_config.hibernation_thresh_vel
-                ):
-                    hibernate = False
-                    break
+    for i_b in range(_B):
+        for island_idx in range(ci.n_islands[i_b]):
+            was_island_hibernated = ci.island_hibernated[island_idx, i_b]
 
-            if hibernate:
-                func_hibernate_entity(
-                    i_e,
-                    i_b,
-                    entities_state=entities_state,
-                    entities_info=entities_info,
-                    dofs_state=dofs_state,
-                    links_state=links_state,
-                    geoms_state=geoms_state,
+            if ti.static(Debug.validate):
+                validate_entity_hibernation_state_for_all_entities_in_temp_island(
+                    island_idx, i_b, entities_state, contact_island, expected_hibernation_state=was_island_hibernated
                 )
-            else:
-                # update collider sort_buffer
-                for i_g in range(entities_info.geom_start[i_e], entities_info.geom_end[i_e]):
-                    collider_state.sort_buffer.value[geoms_state.min_buffer_idx[i_g, i_b], i_b] = geoms_state.aabb_min[
-                        i_g, i_b
-                    ][0]
-                    collider_state.sort_buffer.value[geoms_state.max_buffer_idx[i_g, i_b], i_b] = geoms_state.aabb_max[
-                        i_g, i_b
-                    ][0]
+
+            if not was_island_hibernated:
+                are_all_entities_okay_for_hibernation = True
+                entity_ref_range = ci.island_entity[island_idx, i_b]
+                for i in range(entity_ref_range.n):
+                    entity_ref = entity_ref_range.start + i
+                    entity_idx = ci.entity_id[entity_ref, i_b]
+
+                    is_entity_fixed = entities_info.n_dofs[entity_idx] == 0
```

Comment:
There is no "better" way to check for fixed entities? Like using some boolean flag?